File: app/src/main/java/com/alvarlagerlof/koda/Projects/ProjectsAdapter.java
package com.alvarlagerlof.koda.Projects;

/**
 * Created by alvar on 2016-07-02.
 */

import android.content.Context;
import android.content.Intent;
import android.preference.PreferenceManager;
import android.support.v4.app.FragmentManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.alvarlagerlof.koda.Editor.EditorActivity;
import com.alvarlagerlof.koda.RemoteConfigValues;
import com.alvarlagerlof.koda.R;
import com.alvarlagerlof.koda.Universal.UniversalLoadingObject;
import com.alvarlagerlof.koda.Universal.UniversalLodingViewHolder;

import java.util.ArrayList;

/**
 * Created by alvar on 2015-07-10.
 */


class ProjectsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {


    private FragmentManager fragmentManager;
    private ArrayList<Object> dataset;
    static final int TYPE_LOADING  = 0;
    static final int TYPE_ITEM     = 1;
    static final int TYPE_NO_ITEMS = 2;



    ProjectsAdapter(ArrayList<Object> dataset, FragmentManager fragmentManager) {
        this.dataset = dataset;
        this.fragmentManager = fragmentManager;
    }


    private static class ViewHolderItem extends RecyclerView.ViewHolder {
        TextView title;
        TextView meta;
        LinearLayout more;

        ViewHolderItem(View itemView) {
            super(itemView);
            title = (TextView) itemView.findViewById(R.id.title);
            meta = (TextView) itemView.findViewById(R.id.meta);
            more = (LinearLayout) itemView.findViewById(R.id.more);
        }

    }

    private static class ViewHolderNoItems extends RecyclerView.ViewHolder {
        ViewHolderNoItems(View itemView) {
            super(itemView);
        }
    }




    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        View loadingView = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_loading, viewGroup, false);
        View itemView = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.projects_item, viewGroup, false);
        View noItemsView = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.projects_item_no_items, viewGroup, false);

        switch (i) {
            case TYPE_LOADING:
                return new UniversalLodingViewHolder(loadingView);
            case TYPE_ITEM:
                return new ViewHolderItem(itemView);
            case TYPE_NO_ITEMS:
                return new ViewHolderNoItems(noItemsView);
        }

        throw new RuntimeException("there is no type that matches the type " + i + " + make sure your using types correctly");
    }


    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int position) {

       if (holder instanceof ViewHolderItem) {
            final Context context = ((ViewHolderItem) holder).title.getContext();

            final ProjectsObjectData object = (ProjectsObjectData) dataset.get(position);


            ((ViewHolderItem) holder).title.setText(object.title);
            ((ViewHolderItem) holder).meta.setText(object.updated);


            ((ViewHolderItem) holder).itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(context, EditorActivity.class);
                    intent.putExtra("privateID", object.privateID);
                    intent.putExtra("publicID", object.publicID);
                    intent.putExtra("title", object.title);
                    context.startActivity(intent);
                }
            });


            ((ViewHolderItem) holder).more.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    ProjectsBottomSheetFragment bottomSheetFragment = new ProjectsBottomSheetFragment();
                    bottomSheetFragment.passData(fragmentManager,
                            object.privateID,
                            object.publicID,
                            object.title,
                            PreferenceManager.getDefaultSharedPreferences(context).getString(RemoteConfigValues.PREF_NICK, ""));

                    bottomSheetFragment.show(fragmentManager, bottomSheetFragment.getTag());

                }
            });



        }
    }



    @Override
    public int getItemViewType(int position) {

        if (dataset.get(position) instanceof UniversalLoadingObject) {
            return TYPE_LOADING;
        }

        if (dataset.get(position) instanceof ProjectsObjectNoItems) {
            return TYPE_NO_ITEMS;
        }

        if (dataset.get(position) instanceof ProjectsObjectData) {
            return TYPE_ITEM;
        }

        return TYPE_LOADING;

    }

    @Override
    public int getItemCount() {
        return dataset.size();
    }

}

File: app/src/main/java/com/alvarlagerlof/koda/Projects/ProjectsAdapter.kt
package com.alvarlagerlof.koda.Projects

/**
 * Created by alvar on 2016-07-02.
 */

import android.content.Intent
import android.preference.PreferenceManager
import android.support.v4.app.FragmentManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.alvarlagerlof.koda.Editor.EditorActivity
import com.alvarlagerlof.koda.R
import com.alvarlagerlof.koda.Vars
import com.alvarlagerlof.koda.Universal.UniversalLoadingObject
import com.alvarlagerlof.koda.Universal.UniversalLodingViewHolder
import java.util.*

/**
 * Created by alvar on 2015-07-10.
 */


internal class ProjectsAdapter(private val data: ArrayList<Any>, private val fragmentManager: FragmentManager) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {


    private class ViewHolderItem internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal var title: TextView = itemView.findViewById(R.id.title_input)
        internal var meta: TextView = itemView.findViewById(R.id.meta)
        internal var more: LinearLayout = itemView.findViewById(R.id.more)
    }

    private class ViewHolderNoItems internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView)



    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): RecyclerView.ViewHolder {
        val loadingView = LayoutInflater.from(viewGroup.context).inflate(R.layout.item_loading, viewGroup, false)
        val itemView = LayoutInflater.from(viewGroup.context).inflate(R.layout.projects_item, viewGroup, false)
        val noItemsView = LayoutInflater.from(viewGroup.context).inflate(R.layout.projects_item_no_items, viewGroup, false)

        when (i) {
            TYPE_LOADING -> return UniversalLodingViewHolder(loadingView)
            TYPE_ITEM -> return ViewHolderItem(itemView)
            TYPE_NO_ITEMS -> return ViewHolderNoItems(noItemsView)
        }

        throw RuntimeException("there is no type that matches the type $i + make sure your using types correctly")
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        if (holder is ViewHolderItem) {
            val context = holder.title.context

            val (privateID, publicID, title, updated) = data[position] as ProjectsObjectData


            holder.title.text = title
            holder.meta.text = updated


            holder.itemView.setOnClickListener {
                val intent = Intent(context, EditorActivity::class.java)
                intent.putExtra("privateID", privateID)
                context.startActivity(intent)
            }


            holder.more.setOnClickListener {
                val bottomSheetFragment = ProjectsBottomSheetFragment(
                        privateID,
                        publicID,
                        title,
                        PreferenceManager.getDefaultSharedPreferences(context).getString(Vars.PREF_NICK, ""))

                bottomSheetFragment.show(fragmentManager, bottomSheetFragment.tag)
            }


        }
    }


    override fun getItemViewType(position: Int): Int {
        return when (data[position]) {
            is UniversalLoadingObject -> TYPE_LOADING
            is ProjectsObjectNoItems -> TYPE_NO_ITEMS
            is ProjectsObjectData -> TYPE_ITEM
            else -> TYPE_LOADING
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    companion object {
        val TYPE_LOADING = 0
        val TYPE_ITEM = 1
        val TYPE_NO_ITEMS = 2
    }

}
